rosidl_runtime_c/src/type_hash.c

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string.h>
+
+#include "rosidl_runtime_c/type_hash.h"
+
+#include "rcutils/error_handling.h"
+
+static const char RIHS01_PREFIX[] = "RIHS01_";
+// Hash representation is hex string, two characters per byte
+static const size_t RIHS_VERSION_IDX = 4;
+static const size_t RIHS_PREFIX_LEN = 7;
+static const size_t RIHS01_STRING_LEN = 71;  // RIHS_PREFIX_LEN + (ROSIDL_TYPE_HASH_SIZE * 2);
+
+static bool _ishexdigit(char c)
+{
+  return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
+}
+
+/// Translate a single character hex digit to a nibble
+static uint8_t _xatoi(char c)
+{
+  if (c >= '0' && c <= '9') {
+    return c - '0';
+  }
+  if (c >= 'A' && c <= 'F') {
+    return c - 'A' + 0xa;
+  }
+  if (c >= 'a' && c <= 'f') {
+    return c - 'a' + 0xa;
+  }
+  return -1;
+}
+
+/// Tranlate a byte value to two hex characters
+static void _xitoa(uint8_t val, char * dest)
+{
+  uint8_t nibble = 0;
+  for (size_t n = 0; n < 2; n++) {
+    nibble = (val >> (4 * (1 - n))) & 0x0f;
+    if (nibble < 0xa) {
+      dest[n] = '0' + nibble;
+    } else {  // 0xa <= nibble < 0x10
+      dest[n] = 'a' + (nibble - 0xa);
+    }
```

Comment:
I have to say I don't love this; it just sort of assumes that there are 2 bytes already allocated in dest.  Could we maybe make this a char dest[2], or alternatively return a small struct with the two values?

Reply:
If I manually inline this into the translation would you feel better about it? That's the only place it's used. I externalized it for what I thought was readability, but contained it to this translation unit.